### ISSUE TITLE
Parameters for registered language providers

### ISSUE THREAD

**Issue by radeksimko** [microsoft/vscode-languageserver-node#785]
Hi,
I'm trying to implement reference count as code lens (as discussed in https://github.com/microsoft/vscode-languageserver-node/issues/495) and trying to use the client directly when calling the server, as discussed in https://github.com/microsoft/vscode-languageserver-node/issues/642

Here is a prototype:

```ts
vscode.commands.registerCommand('client.showReferences', async (position: vscode.Position, context: vscode.ReferenceContext, token: vscode.CancellationToken) => {
	const wsFolder = vscode.workspace.workspaceFolders[0];
	const client = getDocumentClient(wsFolder.uri);

	const doc = vscode.window.activeTextEditor.document;

	const provider = client.client.getFeature(ReferencesRequest.method).getProvider(doc);
	const locations = await provider.provideReferences(doc, position, context, token);

	await vscode.commands.executeCommand('editor.action.showReferences', doc.uri, position, locations);
}),
```

The request is clearly being called, but the parameters seem wrong when compared to how the relevant RPC method is supposed to be called per spec ([`textDocument/references`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_references)):

```json
[
    {
        "textDocument": {
            "uri": "file:///private/var/workspace/tf-test/github/main.tf"
        },
        "position": {
            "line": 0,
            "character": 0
        },
        "context": {
            "includeDeclaration": true
        }
    },
    null
]
```

Specifically the correct request structure is embedded in an array with cancellation token, which doesn't seem relevant to the server anyway.

What I would expect to see is this:

```json
    {
        "textDocument": {
            "uri": "file:///private/var/workspace/tf-test/github/main.tf"
        },
        "position": {
            "line": 0,
            "character": 0
        },
        "context": {
            "includeDeclaration": true
        }
    }
```

In other words I would expect the `token` to be omitted from here:
https://github.com/microsoft/vscode-languageserver-node/blob/092c2afc3ad7e4d2b03fe8ac0deb418ec4276915/client/src/common/client.ts#L1842

Is this a bug?

**Comment by rcjsuen** [comment]
> In other words I would expect the `token` to be omitted from here:
> https://github.com/microsoft/vscode-languageserver-node/blob/092c2afc3ad7e4d2b03fe8ac0deb418ec4276915/client/src/common/client.ts#L1842
> 
> Is this a bug?

1. Can you try omitting `token` by modifying the `client.js` (presumably inside your `node_modules/` folder? What happens then?
2. If you add a breakpoint and step into the code, where does it lead and what can you find?
3. There are `client.sendRequest(*)` calls everywhere that includes a `token` as its last parameter so this feels like it's either a bug that affects all requests or is working correctly.

**Comment by radeksimko** [comment]
> Can you try omitting token by modifying the client.js (presumably inside your node_modules/ folder? What happens then?

Yep, I did try that and it works as expected - at least the request is received by the server, parsed correctly and responded to, as expected.

> There are client.sendRequest(*) calls everywhere that includes a token as its last parameter so this feels like it's either a bug that affects all requests or is working correctly.

Yes, that's the same conclusion I came to, although it seemed unlikely at first? That's why I raised this issue.

**Comment by rcjsuen** [comment]
> > There are client.sendRequest(*) calls everywhere that includes a token as its last parameter so this feels like it's either a bug that affects all requests or is working correctly.
> 
> Yes, that's the same conclusion I came to, although it seemed unlikely at first? That's why I raised this issue.

That's fair. What did you find while stepping through the code?

**Comment by rcjsuen** [comment]
https://github.com/microsoft/vscode-languageserver-node/blob/cf5d1e4d46efd85af4812b9b5dda775067c753fa/jsonrpc/src/common/connection.ts#L1254-L1259

From my own testing, the code gets to here and `messageParams` looks fine to me.

```JSON
{
    "textDocument": {
        "uri": "file:///a/b/c/d.txt"
    },
    "position": {
        "line": 1,
        "character": 6
    },
    "context": {
        "includeDeclaration": true
    }
}
```

**Comment by radeksimko** [comment]
That is not what I see in my testing - I added the following log point:

```ts
console.log(`${requestMessage.id} - ${requestMessage.method}:`); console.log(requestMessage.params);
```

right after the `requestMessage` is set and I can clearly see that an array is being passed:

![Screenshot 2021-07-12 at 09 26 09](https://user-images.githubusercontent.com/287584/125255371-3fc90100-e2f3-11eb-91de-104bee6286c4.png)


FWIW this is in `node_modules/vscode-jsonrpc/lib/common/connection.js` and `vscode-languageclient` is at `^8.0.0-next.1`

**Comment by rcjsuen** [comment]
> That is not what I see in my testing

@radeksimko Interesting that you now have `null` instead of `token-example`.

`messageParams` is defined in that `sendRequest` function so maybe you can step through the code slowly to see what's going on.

**Comment by radeksimko** [comment]
Oh, sorry for the added confusion - the `token-example` in my very first comment was _literally_ an example where I statically passed a string as a token from the (slightly modified) prototype code.

FWIW my TypeScript skills are limited and I'm not even sure what's the role of the cancellation token in that context there to be honest - I'm literally trying to just satisfy the interface 🤷🏻‍♂️  The server probably isn't one that should be generating any kind of token as it isn't one that would be cancelling that request that way anyway (server-side cancellation is done via separate method per LSP spec).

> messageParams is defined in that sendRequest function so maybe you can step through the code slowly to see what's going on.

I can see what's going on, but I don't know what is _supposed_ to be going on - especially with regards to the token.

AFAICT the token should probably only be passed to the failure handler, but not to `sendRequest` at all?

https://github.com/microsoft/vscode-languageserver-node/blob/092c2afc3ad7e4d2b03fe8ac0deb418ec4276915/client/src/common/client.ts#L1841-L1847

I am guessing that the root cause is that there is many `sendRequest` implementations with different arguments and the last one is being matched, when the original implementation was built around the others (which accept cancellation token):

https://github.com/microsoft/vscode-languageserver-node/blob/092c2afc3ad7e4d2b03fe8ac0deb418ec4276915/client/src/common/client.ts#L85-L91

**Comment by rcjsuen** [comment]
https://github.com/microsoft/vscode-languageserver-node/blob/092c2afc3ad7e4d2b03fe8ac0deb418ec4276915/client/src/common/client.ts#L1839-L1841

When you get here, what do your four parameters of `document`, `position`, `options`, and `token` look like?

https://github.com/microsoft/vscode-languageserver-node/blob/cf5d1e4d46efd85af4812b9b5dda775067c753fa/jsonrpc/src/common/connection.ts#L1193-L1195

Then when it gets down to here, what does `type` and `args` look like?

**Comment by dbaeumer** [comment]
Regarding the original description. Passing the token is correct. The json rpc library takes the token and hooks listeners to convert a cancellation into a LSP cancelation message. 

For some reason the request is send as byPosition instead of byName.

Can you debugged here: https://github.com/microsoft/vscode-languageserver-node/blob/main/jsonrpc/src/common/connection.ts#L1194 and see why this is happening. See also https://github.com/microsoft/vscode-languageserver-node/blob/main/jsonrpc/src/common/messages.ts#L176

In LSP everything should be sent byName.

**Comment by dbaeumer** [comment]
OK. The problem is that you pass null as a cancellation token. This is then interpreted as an additional argument and hence the parameters are passed a byPosition. Please ensure you provide a proper cancellation token. You can always create one using the following code:

```
	let tokenSource: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
        let token = tokenSource.token;

```

**Comment by radeksimko** [comment]
Thank you, that worked!
